### PR TITLE
executor: fix wrong key range of index scan when filter is comparing year column with NULL

### LINKED ISSUE #23038: wrong result of "left join where year != null"
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
use test;
drop table if exists t;
CREATE TABLE `t` (
  `COL1` year(4) DEFAULT NULL COMMENT 'DATE AND TIME',
  `COL2` varchar(20) DEFAULT NULL,
  KEY `UK_COL1` (`COL1`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin;
insert into t values(1, "1");
insert into t values(1, "2");
select *  from t as t1 left join t as t2 on t1.col1 = t2.col1 where t1.col1 != NULL;
select count(*) from t as t1 left join t as t2 on t1.col1 = t2.col1 where t1.col1 != NULL;
select t1.col1, t2.col2 from t as t1 left join t as t2 on t1.col1 = t2.col1 where t1.col1 != NULL;
```

### 2. What did you expect to see? (Required)
mysql 8.0.23
```
mysql> select *  from t as t1 left join t as t2 on t1.col1 = t2.col1 where t1.col1 != NULL;
Empty set (0.04 sec)

mysql> select count(*) from t as t1 left join t as t2 on t1.col1 = t2.col1 where t1.col1 != NULL;
+----------+
| count(*) |
+----------+
|        0 |
+----------+
1 row in set (0.04 sec)

mysql> select t1.col1, t2.col2 from t as t1 left join t as t2 on t1.col1 = t2.col1 where t1.col1 != NULL;
Empty set (0.04 sec)
```
### 3. What did you see instead (Required)
master
```
mysql> select *  from t as t1 left join t as t2 on t1.col1 = t2.col1 where t1.col1 != NULL;
Empty set (0.00 sec)

mysql> select count(*) from t as t1 left join t as t2 on t1.col1 = t2.col1 where t1.col1 != NULL;
+----------+
| count(*) |
+----------+
|        4 |
+----------+
1 row in set (0.00 sec)

mysql> select t1.col1, t2.col2 from t as t1 left join t as t2 on t1.col1 = t2.col1 where t1.col1 != NULL;
+------+------+
| col1 | col2 |
+------+------+
| 2001 | NULL |
| 2001 | NULL |
+------+------+
2 rows in set (0.00 sec)
```

release-4.0 and release-5.0-rc
```
mysql> select *  from t as t1 left join t as t2 on t1.col1 = t2.col1 where t1.col1 != NULL;
+------+------+------+------+
| COL1 | COL2 | COL1 | COL2 |
+------+------+------+------+
| 2001 | 1    | 2001 | 1    |
| 2001 | 1    | 2001 | 2    |
| 2001 | 2    | 2001 | 1    |
| 2001 | 2    | 2001 | 2    |
+------+------+------+------+
4 rows in set (0.00 sec)

mysql> select count(*) from t as t1 left join t as t2 on t1.col1 = t2.col1 where t1.col1 != NULL;
+----------+
| count(*) |
+----------+
|        4 |
+----------+
1 row in set (0.00 sec)

mysql> select t1.col1, t2.col2 from t as t1 left join t as t2 on t1.col1 = t2.col1 where t1.col1 != NULL;
+------+------+
| col1 | col2 |
+------+------+
| 2001 | 1    |
| 2001 | 2    |
| 2001 | 1    |
| 2001 | 2    |
+------+------+
4 rows in set (0.00 sec)
```
### 4. What is your TiDB version? (Required)

master dc9ceb4bba96608de2aa5ac906e9ad003fa8a46c
release-5.0-rc f41ca49fab41af6ba168e94e6eeeef517778cee1
release-4.0 8406e7f58b1d5811966a6811d7b519bcfd6020b2

### PR DESCRIPTION
**Pull request by guo-shaoge**
Signed-off-by: guo-shaoge <shaoge1994@163.com>

### What problem does this PR solve?

Issue Number: close #23038 

Problem Summary: use following case:

    use test;
    drop table if exists gjttbl;
    CREATE TABLE `gjttbl` (
        `gjtc1` year(4) DEFAULT NULL COMMENT 'DATE AND TIME',
        `gjtc2` varchar(20) DEFAULT NULL,
        KEY (`gjtc1`)
        );
    insert into gjttbl values(1, "1");
    insert into gjttbl values(1, "2");
    select  t1.gjtc1, t2.gjtc1  from gjttbl as t1 inner join gjttbl as t2 on t1.gjtc1 = t2.gjtc1 where t1.gjtc1 != NULL;

Output should be empty, but TiDB give result.
### What is changed and how it works?

What's Changed: change datum.go:convertToMysqlYear(), when arg is NULL, this function accidentally set result tobe zero year.

How it Works: go:convertToMysqlYear() is called by ranger.(*builder).buildFormBinOp(), which will build key range. So if go:convertToMysqlYear() change NULL to zero, the key range for index scan is wrong.

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Integration test: tidb_test.go:TestYear()

Side effects: no

### Release note 

- fix wrong key range of index scan when filter is comparing year column with NULL

### CODE REVIEW

File: types/datum.go [hunk 1]
```diff
@@ -1424,6 +1424,9 @@ func (d *Datum) convertToMysqlFloatYear(sc *stmtctx.StatementContext, target *Fi
 		y = float64(d.GetMysqlTime().Year())
 	case KindMysqlDuration:
 		y = float64(time.Now().Year())
+	case KindNull:
+		ret = *d
```
Comment (XuHuaiyu) [inline types/datum.go:1428]: Add some comment for this modification

File: server/server_test.go [hunk 2]
```diff
@@ -1595,6 +1595,35 @@ func (cli *testServerClient) runTestIssue3682(c *C) {
...
 
+func (cli *testServerClient) runTestIndexScanWithYearCol(c *C) {
+	cli.runTests(c, nil, func(dbt *DBTest) {
+		dbt.mustExec("use test;")
+		dbt.mustExec("drop table if exists t;")
+		dbt.mustExec("create table t (c1 year(4), c2 int, key(c1));")
+		dbt.mustExec("insert into t values(2001, 1);")
+
+		// test inner join
+		rows := dbt.mustQuery("select t1.c1, t2.c1 from t as t1 inner join t as t2 on t1.c1 = t2.c1 where t1.c2 != NULL;")
```
Comment (XuHuaiyu) [inline server/server_test.go:1606]: We also need to test the explain result.
Please move this test to the package executor. You can take the test func `TestInjectProjBelowTopN` as an example

File: types/datum.go (hunk 1 above)
Comment (guo-shaoge) [reply types/datum.go:1429]: done

**Review by XuHuaiyu — COMMENTED** [review]
LGTM

**Comment by AilinKid** [comment]
you clean the reluctant title suffix in the PR description `…year column with NULL`

**Comment by guo-shaoge** [comment]
> you can clean the reluctant title suffix in the PR description `…year column with NULL`

done

**Pull request merged**